esphome/components/modbus/modbus.cpp

```diff
@@ -133,38 +168,78 @@ bool Modbus::parse_modbus_byte_(uint8_t byte) {
     if (device->address_ == address) {
       // Is it an error response?
       if ((function_code & 0x80) == 0x80) {
-        ESP_LOGD(TAG, "Modbus error function code: 0x%X exception: %d", function_code, raw[2]);
-        if (waiting_for_response != 0) {
-          device->on_modbus_error(function_code & 0x7F, raw[2]);
+        uint8_t exception = raw[2];
+        ESP_LOGD(TAG, "Modbus error function code: 0x%X exception: %d", function_code, exception);
+        if (waiting_for_response_ == address) {
+          set_timeout("on_modbus_error", 0, [device, function_code, exception]() {
+            device->on_modbus_error(function_code & 0x7F, exception);
+          });
         } else {
           // Ignore modbus exception not related to a pending command
           ESP_LOGD(TAG, "Ignoring Modbus error - not expecting a response");
         }
       } else if (this->role == ModbusRole::SERVER && (function_code == 0x3 || function_code == 0x4)) {
-        device->on_modbus_read_registers(function_code, uint16_t(data[1]) | (uint16_t(data[0]) << 8),
-                                         uint16_t(data[3]) | (uint16_t(data[2]) << 8));
+        set_timeout("on_modbus_read_registers", 0, [device, data, function_code]() {
+          device->on_modbus_read_registers(function_code, uint16_t(data[1]) | (uint16_t(data[0]) << 8),
+                                           uint16_t(data[3]) | (uint16_t(data[2]) << 8));
+        });
       } else {
-        device->on_modbus_data(data);
+        set_timeout("on_modbus_data", 0, [device, data]() { device->on_modbus_data(data); });
       }
       found = true;
     }
   }
-  waiting_for_response = 0;
 
   if (!found) {
     ESP_LOGW(TAG, "Got Modbus frame from unknown address 0x%02X! ", address);
   }
 
-  // reset buffer
-  ESP_LOGV(TAG, "Clearing buffer of %d bytes - parse succeeded", at);
-  this->rx_buffer_.clear();
+  clear_rx_buffer_("parse succeeded");
+
+  if (waiting_for_response_ == address)
+    waiting_for_response_ = 0;
+
   return true;
 }
 
+void Modbus::send_next_frame_() {
+  if (this->tx_buffer_.empty())
+    return;
+
+  if (tx_blocked())
+    return;
+
+  std::vector<uint8_t> data = this->tx_buffer_.front();
+
+  if (this->role == ModbusRole::CLIENT) {
+    waiting_for_response_ = data[0];
+  }
+
+  if (this->flow_control_pin_ != nullptr)
+    this->flow_control_pin_->digital_write(true);
+
+  this->write_array(data);
+  this->flush();
+
+  if (this->flow_control_pin_ != nullptr)
+    this->flow_control_pin_->digital_write(false);
+
+  this->tx_buffer_.pop_front();
+
+  last_send_ = millis();
```

Comment:
I think this is enough to prevent unnecessary blocking:
```suggestion
  if (this->flow_control_pin_ != nullptr) {
    this->flush();
    this->flow_control_pin_->digital_write(false);
    last_send_ = millis();
  }
  else {
    last_send_ = millis() + (((10 * data.size() * 11 * 1000 / this->parent_->get_baud_rate()) + 5) / 10);
  }
  
  this->tx_buffer_.pop_front();
  
```
The ((10 * x) + 5) / 10) ensures rounding up ;)

Reply:
Something like this could be used to avoid blocking, but the millis() + offset method won't work. 
If we later compare this value to millis() and less than offset ms have elapsed we'll get a negative number (which will wrap to a large positive number) and falsely tell us that we're past millis() + offset when we're not.

I'll propose an alternative that achieves the non-blocking behaviour.